[draft] materialization for FLIP 158

curcur opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -329,37 +330,50 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
+            SnapshotResult<KeyedStateHandle> materialized) {
+
+        // collections don't change once started and handles are immutable
+        List<ChangelogStateHandle> prevDeltaCopy =
+                new ArrayList<>(periodicMaterializer.restoredNonMaterialized());
```
I'm afraid this access needs synchronization:
- `periodicMaterializer.restoredNonMaterialized` is written by the task thread
- this method `buildSnapshotResult()` can be called by the DFS writer thread

curcur commented:
TODO: tests

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(materializedState.lastMaterializedTo())) {
+                        return;
+                    }
```
I think we should reset `materializationOnGoing` flag here.

curcur replied:
you are right!

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
```
Should we handle any possible exceptions here to honour retry limit?

curcur replied:
Do you mean the sync phase? In the sync phase, if it fails, the task should fail?

Or do we want to allow sync phase to fail as well? Can you give me an example that the sync phase fails but we should tolerate that?

rkhachatryan replied:
Yes, I mean the sync phase. I think recoverable I/O failures are possible here, but I'm not sure if they should be tolerated TBH.

For example: timeout writing to SAN/NAS, no disk space, file handle limit exceeded, bad sectors not handled by the FS.

If it's too difficult to honour the limit here then failover is also an option.

curcur replied:
The problem is that most of the sync failure is non-tolerable/recoverable (or do not know whether we can tolerate or recovered). It is difficult to distinguish all these cases.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    materializedId++, // TO DISCUSS: seems this does not matter,
```
Could you add the results of the last discussion as a comment here?
(i.e. some backends might need use `id`; however it doesn't make sense for materialization; we should probably add a new method or store the id)

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture,
+            SequenceNumber upTo) {
+
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
```
Can we close the safety net right after this statement (`runIfNotDoneAndGet`)?
I think it would make the flow easier to follow and possible exceptions won't affect each other.

curcur replied:
but when exceptions happen, we also need to close the safety net, that's why put within `finally`

rkhachatryan replied:
`mailboxExecutor.execute` and everything that's inside doesn't need this.
Then maybe split `asyncMaterialization` function into two:
  - `materialize` (try/finally with safety net)
  - `mailboxExecutor.execute` (and if result is `null`/empty then only update `materializationOnGoing` - in the same mail)
  ?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+        FileSystemSafetyNet.initializeSafetyNetForThread();
+        try {
+            FutureUtils.runIfNotDoneAndGet(materializedRunnableFuture);
+            mailboxExecutor.execute(
+                    () -> {
+                        materializedState =
+                                new MaterializedState(
+                                        getMaterializedResult(materializedRunnableFuture.get()),
```
I think it's safer to call `future.get` from the async thread (even if it's a no-op currently).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+                    retryTime,
+                    e);
+
+            // ToDO: Double check
+            //  how the async phase is related to materialized state tracking
+            //  refer to AsyncCheckpointRunnable#run#catch
+            //  cleanup logic in StateUtil#discardStateFuture
```
This should be fine (for tracking we need a Tracker anyway, and I think it should not affect this part of code).

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+            //  cleanup logic in StateUtil#discardStateFuture
+            handleExecutionException(materializedRunnableFuture);
+
+            if (retryTime == allowedNumberOfFailures) {
+                asyncExceptionHandler.handleAsyncException(
+                        "Fail to complete the asynchronous part of materialization", e);
+            }
```
1. I think we should either reset failure counter here or check it before each materialization.. Otherwise, we depend on asyncExceptionHandler for consistent behaviour.
2. Replace `==` with `>=`?

curcur replied:
>=; done!

what do you mean by `asyncExceptionHandler` consistent behavior? why should we reset to 0 if we want to fail reaching max retry number?

Why do I have to check the max number before materialization (sync phase I guess you mean) before it fails? Doesn't it can bump up only if a failure happens?

rkhachatryan replied:
Maybe let's discuss first what should happen with subsequent materialization once `numberOfConsecutiveFailures` reaches the limit?

If we want them to run, should they have the same (configured) number of attempts? Then we should reset it here?
If we don't, i.e. we prevent any further materializations, then we should check this counter before starting materialization?

Currently, the latter option I think is enforced implicitly by `asyncExceptionHandler` - which fails the job.

curcur replied:
The contract is:,

If reaching the limit (unrecoverable), we should fail the task.

`asyncExceptionHandler` is a way to fail the task.

resetting te counter is misleading and confusing to this contract.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+
+    private final CheckpointStreamFactory streamFactory;
+
+    private final Snapshotable keyedStateBackend;
```
nit: rename to `snapshotable`?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+    private final int allowedNumberOfFailures;
+
+    /** Materialization failure retries. */
+    private final AtomicInteger retries;
```
Rename to `numberOfFailures`/`numberOfConsecutiveFailures`?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+    private final AtomicInteger retries;
+
+    /** Making sure only one materialization on going at a time. */
+    private final AtomicBoolean materializationOnGoing;
```
I think this can be boolean (non-atomic) - as it's only accessed only a single (task) thread.

curcur replied:
yes, I was thinking back-and-force whether this flag to only be touched in the task thread. For now, it can be non-atomic.

curcur replied:
Let's keep it for now (I think it may also be changed later, if we want to trigger on completion or use pause). Let me refine this after making the final decision.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+        this.keyedStateBackend = keyedStateBackend;
+        this.stateChangelogWriter = stateChangelogWriter;
+        this.asyncExceptionHandler = asyncExceptionHandler;
+        this.periodicExecutor =
+                Executors.newSingleThreadScheduledExecutor(
+                        new ExecutorThreadFactory("periodic-materialization"));
```
nit: add `-scheduler` to name? and ideally task info (`env.getTaskInfo().getTaskNameWithSubtasks()`)

curcur replied:
Yes, I think add task name is a good idea!

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -329,37 +332,47 @@ public boolean deregisterKeySelectionListener(KeySelectionListener<K> listener)
-                        new ChangelogStateBackendHandleImpl(
-                                materialized, prevDeltaCopy, getKeyGroupRange()));
-            }
+    @Override
+    @VisibleForTesting
+    public void triggerMaterialization() {
+        periodicMaterializer.triggerMaterialization();
+    }
```
This is very..practical :slightly_smiling_face: 
I think a similar effect can be achieved by using some custom `ScheduledExecutorService` in Materializer (e.g. `ManuallyTriggeredScheduledExecutorService`)

curcur replied:
:-)

curcur replied:
I've played with `ManuallyTriggeredScheduledExecutorService` a bit
It does not make things cleaner or less exposure of methods/APIs, but introducing more complexities for PeriodicMaterializer constructor (now I have to expose a way to substitute the executor).

The reason here is I want to test materialization itself works correctly. To do that, I have to expose the "triggerMaterialization" method, and trigger it through the `ManuallyTriggeredScheduledExecutorService` manually.

so "triggerMaterialization" has to be exposed anyways. 

Please let me knwo whether this makes sense to you, or if you have better suggestions.

rkhachatryan replied:
I tried to disable materialization by commenting it in `ChangelogKeyedStateBackend.triggerMaterialization` and the test still passed, so I think just triggering is not enough.

This suggests, that if we want to test it on this level, a more fine-grained control over materialization is needed; which can be gained by a direct access to the materializer in test.

And if we have such access, exposing something through the backend isn't necessary.
But probably it makes sense to first resolve the questions about backend-materializer interaction (like [this](https://github.com/apache/flink/pull/16606#discussion_r682621905) one).

WDYT?

curcur replied:
> I tried to disable materialization by commenting it in `ChangelogKeyedStateBackend.triggerMaterialization` and the test still passed, so I think just triggering is not enough.

That's expected, that's equivalent to disable materialization that uses purely changelog.

rkhachatryan replied:
As we discussed offline, this only works because a direct executor is used when creating materilizier (by "works" I mean the call waits for the materialization).
Please add a comment at the very least.

However, I still think the test or production code structure is wrong. With a proper separation between materializer and backend, we should be able to listen for materialization completion and take a snapshot only after that (instead of relying on the internal materializer behavior).

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java:
```diff
@@ -345,7 +353,12 @@ public ExecutorService getAsyncOperationsThreadPool() {
     @Override
-    public ThroughputCalculator getThroughputMeter() {
-        return throughputCalculator;
+    public void setCheckpointStorage(CheckpointStorage checkpointStorage) {
+        this.checkpointStorage = checkpointStorage;
+    }
```
Could we add a check here to prevent resetting the storage?
And also to other recently added setters?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/RuntimeEnvironment.java:
```diff
@@ -316,6 +319,11 @@ public TaskOperatorEventGateway getOperatorCoordinatorEventGateway() {
         return operatorEventGateway;
     }
 
+    @Override
+    public ThroughputCalculator getThroughputMeter() {
+        return throughputCalculator;
+    }
+
```
nit: unrelated change

curcur replied:
yes, I moved all "final" attributes all together along with their get method for easy reading. 

I can maybe create a hotfix on this if you inisit.

rkhachatryan replied:
I see, I think it's fine to use the same commit.

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:
```diff
@@ -127,13 +128,29 @@
 
     private boolean isLatencyTrackingConfigured = false;
 
+    /** Interval in milliseconds to perform periodic materialization. */
+    private long periodicMaterializeInterval =
+            StateBackendOptions.PERIODIC_MATERIALIZATION_INTERVAL.defaultValue();
```
nit: can we add `Millis` or `Ms` to the variable name (as in `taskCancellationIntervalMillis`)?
ditto other added fields

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:
```diff
@@ -127,13 +128,29 @@
+    private long periodicMaterializeInterval =
+            StateBackendOptions.PERIODIC_MATERIALIZATION_INTERVAL.defaultValue();
+
+    /** Interval in milliseconds for initial delay of periodic materialization. */
+    private long periodicMaterializeInitDelay =
+            StateBackendOptions.PERIODIC_MATERIALIZATION_INIT_DELAY.defaultValue();
```
Why `periodicMaterializeInterval` isn't sufficient?

curcur replied:
That's the init delay. I can remove it if you think it is not necessary. I do not have a strong opinion.

rkhachatryan replied:
I think the interval should be enough (one less option to learn for users).

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:
```diff
@@ -127,13 +128,29 @@
+    private long periodicMaterializeInitDelay =
+            StateBackendOptions.PERIODIC_MATERIALIZATION_INIT_DELAY.defaultValue();
+
+    /** Max allowed number of failures */
+    private int materializationMaxAllowedFailures =
+            StateBackendOptions.MATERIALIZATION_MAX_ALLOWED_FAILURES.defaultValue();
```
nit: the comment says the same as the variable name so it can be removed
ditto other added fields

curcur replied:
Make the comment more informative.

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java:
```diff
@@ -127,13 +128,29 @@
+    private int materializationMaxAllowedFailures =
+            StateBackendOptions.MATERIALIZATION_MAX_ALLOWED_FAILURES.defaultValue();
+
+    /** Flag to enable periodic materialization */
+    private boolean isPeriodicMaterializationEnabled = false;
```
What are the use caseswhen this flag is false?

curcur replied:
I was thinking for testing purposes. 

Not for the unit test, but for the randomization test, when we only want to test changelog itself, not including materialization.

WDYT?

rkhachatryan replied:
I'm not sure whether we should randomize it; but it seems we can achieve it with a very big materialization interval, can't we?

curcur replied:
Not to randomize materialization. It provides a way that only uses changelog  
OR changelog + materialization.

For testing purposes (at least for now), I think it is quite important.

Using very big intervals could also be a solution, but it depends on the implementation of how materialization is implemented, which as you commented (very practical).

rkhachatryan replied:
Yes, I got what you meant regarding "changelog OR changelog + materialization" and that's why I proposed an alternative (big intervals).
It doesn't introduce new config option (and new code) and that's why I think it's preferrable.

rkhachatryan replied:
Resolved by removing the flag.

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/StateBackendOptions.java:
```diff
@@ -88,4 +88,35 @@
                     .withDescription(
                             "Defines the number of measured latencies to maintain at each state access operation.");
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_MATERIALIZATION)
+    public static final ConfigOption<Long> PERIODIC_MATERIALIZATION_INTERVAL =
+            ConfigOptions.key("state.backend.periodic-materialize.interval")
+                    .longType()
+                    .defaultValue(10000L)
+                    .withDescription(
+                            "Defines the interval in milliseconds to perform periodic materialization for state backend.");
```
I think these options should be placed together with other `state.backend.changelog*` options (`CheckpointingOptions`) and use the same `COMMON_STATE_BACKENDS` section.
WDYT?

curcur replied:
I am thinking to create a separate ChangelogOptions class, for all such configurations.

curcur replied:
Moved all changelog related configurations to `ChangelogOptions`, using "CHANGELOG" session?

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/StateBackendOptions.java:
```diff
@@ -88,4 +88,35 @@
+                            "Defines the interval in milliseconds to perform periodic materialization for state backend.");
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_MATERIALIZATION)
+    public static final ConfigOption<Long> PERIODIC_MATERIALIZATION_INIT_DELAY =
+            ConfigOptions.key("state.backend.periodic-materialize.initial-delay")
+                    .longType()
+                    .defaultValue(10000L)
```
I think this default is too small if we are targeting 500-1000ms checkpoint duration.
"Continuous materialization" is also an option but maybe we should start with the default of `execution.checkpointing.timeout` - 10min?

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/StateBackendOptions.java:
```diff
@@ -88,4 +88,35 @@
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_MATERIALIZATION)
+    public static final ConfigOption<Long> PERIODIC_MATERIALIZATION_INTERVAL =
+            ConfigOptions.key("state.backend.periodic-materialize.interval")
+                    .longType()
+                    .defaultValue(10000L)
```
What about `durationType`?
ditto other added fields

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/AsyncExceptionHandler.java:
```diff
@@ -0,0 +1,34 @@
+import org.apache.flink.runtime.execution.Environment;
+
+/** Handling exceptions thrown by non-task thread. */
+public class AsyncExceptionHandler {
```
This class looks like a copy of `StreamTaskAsyncExceptionHandler` - why didn't you use the existing one (plus interface)?

ditto `AsynchronousException`

curcur replied:
Yes, should reuse it.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/MaterializedState.java:
```diff
@@ -0,0 +1,61 @@
+            List<KeyedStateHandle> materializedSnapshot,
+            List<ChangelogStateHandle> restoredNonMaterialized,
+            SequenceNumber materializedTo) {
+        this.materializedSnapshot = materializedSnapshot;
+        this.restoredNonMaterialized = restoredNonMaterialized;
+        this.materializedTo = materializedTo;
```
`notNull` check and wrap with `unmodifiable`?

curcur replied:
good point!

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogStateBackend.java:
```diff
@@ -222,8 +223,11 @@ public StateBackend configure(ReadableConfig config, ClassLoader classLoader)
                                 env.getMainMailboxExecutor(),
-                                env.getAsyncOperationsThreadPool()));
+                                env.getAsyncOperationsThreadPool(),
+                                env.getCheckpointStorage()
+                                        .createCheckpointStorage(env.getJobID())));
```
I think it's better to pass `CheckpointStorageAccess` created in `StreamTask` as an argument, because:
- The returned object is supposed to be used by multiple backends so it doesn't seem like a backend responsibility to create it
- the call can also potentially create a new storage. 
- if it has to be closed then it will be unclear who is doing it (existing backends using it are deprecated) 

(at least add a todo)

curcur replied:
After reading the code, I agree that we should follow the current contract that sharing the same `CheckpointStorageAccess`, so that the ownership still kept in `StreamTask`

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/test/java/org/apache/flink/contrib/streaming/state/EmbeddedRocksDBStateBackendTest.java:
```diff
@@ -610,6 +612,10 @@ public void testSharedIncrementalStateDeRegistration() throws Exception {
         }
     }
 
+    @Override
+    @Test
+    public void testMaterializedRestore() {}
+
```
This override prevents the test from running at all, even for changelog backend tests.

curcur replied:
The delegated state backend does not have materialization; that's why the test is ignored.

rkhachatryan replied:
Yes, but with this override even the changelog backend isn't tested: its tests inherit this empty method.

curcur replied:
LOL, that's true! Thanks for point this out!

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+    }
+
+    // TODO: this method may change after the ownership PR
+    public static List<KeyedStateHandle> getMaterializedResult(
```
This can be private.

rkhachatryan reviewed (changes requested):
Thanks for the PR @curcur,
I have left some comments, PTAL.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+                                        upTo);
+
+                        checkState(
+                                materializationOnGoing.compareAndSet(true, false),
```
That's a good idea, info level should also be fine I think.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** Periodically make materialization for the delegated state backend. */
+public class PeriodicMaterializer {
```
I am actually thinking `PeriodicMaterializer` a good name, it is clear what it is and how it achieve this.

Maybe later we will have different Materialzier as well.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // Only one materialization ongoing at a time
+                    if (!materializationOnGoing.compareAndSet(false, true)) {
+                        return;
+                    }
```
That's a good point!

rkhachatryan replied:
Triggering immediately seems tricky to implement because we need to wait for the async phase to finish. 
So I'd go with min-pause solution and sceduling the next materialization upon finishing the current one. Let's discuss it offline.

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,263 @@
+                        return;
+                    }
+
+                    SequenceNumber upTo = stateChangelogWriter.lastAppendedSequenceNumber().next();
```
NON-materialized is not part of the materialized state, that part I agree.
However, I do think NON-materialized + materialized state needs to be put together.

How about this: put NON-materialized, materialized state and seq number together as ChangelogKeyedStateBackend's state, and provides API to access this state from PeriodicMaterializer.

curcur replied:
The reason is that I still want to keep all the materialization (sync and async phase within the materializer), and do not want to expose the details.

The only interaction between Keyedstatebackend and materializer is Keyedstatebackend `state` update

rkhachatryan replied:
> How about this: put NON-materialized, materialized state and seq number together as ChangelogKeyedStateBackend's state, and provides API to access this state from PeriodicMaterializer.

This is similar to what I proposed above, except that you want materializer, not the backend run the sync phase, right?
I guess the motivation is to have "materialization" in code covering both sync and async phases?

This may be reasonable, but OTH we lose binding between SQN and sync-snapshot which makes it more error-prone.
Besides that, the final solution seems cleaner if sync-phase runs in backend, as all mailbox-requiring activities may reside in one class.

curcur replied:
1. the sqn and sync-snapshot are still bonded together in my current solution

2. I want to sync + async to be wrapped in the materializer; the backend is only responsible to provide a way to make a snapshot.

Changelog statebackend contains logic -> materilizer + changed log 

materializer -> sync + async

ideally, I think changed log is better to be wrapped as well, but I am fine with it right now.

rkhachatryan replied:
Resolved by using the proposal [above](https://github.com/apache/flink/pull/16606#discussion_r682621905).

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/materializer/MaterializedState.java:
```diff
@@ -0,0 +1,61 @@
+    /** Set initially on restore and later upon materialization. */
+    private final List<KeyedStateHandle> materializedSnapshot;
+
+    /** Updated initially on restore and later cleared upon materialization. */
+    private final List<ChangelogStateHandle> restoredNonMaterialized;
```
see [above](https://github.com/apache/flink/pull/16606#discussion_r691055101)

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializer.java:
```diff
@@ -0,0 +1,260 @@
+    public void triggerMaterialization() {
+        mailboxExecutor.execute(
+                () -> {
+                    // synchronize phase
+                    SequenceNumber upTo =
+                            stateChangelogWriter.lastAppendedSequenceNumber().next();
+
+                    if (upTo.equals(changelogStateBackendState.lastMaterializedTo())) {
+
+                        scheduleNextMaterialization();
+
+                        LOG.info(
+                                "Task {} has no state updates since last materialization, "
+                                        + "skip this one and schedule the next one in {} seconds",
+                                subtaskName,
+                                periodicMaterializeDelay / 1000);
+
+                        return;
+                    }
+
+                    RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture =
+                            keyedStateBackend.snapshot(
+                                    // This ID is not needed for materialization;
+                                    // But since we are re-using the streamFactory
+                                    // that is designed for state backend snapshot,
+                                    // which requires unique checkpoint ID.
+                                    // A faked materialized Id is provided here.
+                                    // TODO: implement its own streamFactory.
+                                    materializedId++,
+                                    System.currentTimeMillis(),
+                                    streamFactory,
+                                    CHECKPOINT_OPTIONS);
```
logically unclear if moving this part (sycn part) to changelog statebakcned.
In that case, schedule happens both in materializer and changelogstatebackend.

curcur commented:
Roman and I had several long discussions on interfaces between `Materialization` and `ChangelogKeyedStatebackend`. Document here for future reference.

The main difference is between who is responsible to **keep and update** the ChangelogKeyedStatebackend related states, denoted as `ChangelogSnapshotState`, including three parts:

 * - materialized snapshot from the underlying delegated state backend
 * - non-materialized part in the current changelog
 * - non-materialized changelog, from previous logs (before failover or rescaling) 

We've discussed and tried out three versions:

1. `Materialization` coupled with `ChangelogKeyedStatebackend`, 
implemented in commit **fbd1e2d38ae6353506ceac8eb074bd24bdb29b62**
	Where `PeriodicMaterializer` is an inner class of `ChangelogKeyedStatebackend`
	- Pros: states are shared, easy to reason about
	- Cons: Coupled too closely, not flexible or extendible for keyedstatebackend or materializer

	Not to mention further, this approach is discarded during early discussion.
	
2. `ChangelogSnapshotState` are kept in materializer. Materializer is conceptually taken as a way to connect delegated state backend to changelog. How to connect: through `ChangelogSnapshotState`, as denoted above.
implemented in commit **3421b81c2502f61112bd131a7336c16e3dd30925**

    - Pros: 
      1. Good isolation and extensibility. Clear view the changelog keyedstatebackend as four parts: 
	    - log writer, delegated statebackend, materializer, and wrapper changelogkeyedstatebackend for double writing
      2. More natural to understand and implement.
	    - State is updated by the materializer, and accessible by changelogKeyedStateBackend
	    - Materializer is part of ChangelogKeyedStateBackend

    - Cons: 
	  1. according to Roman, ChangelogKeyedStateBackend has implicit states (like state double writes) besides the three mentioned above; 
	  2. optimization (like batched writes) need to update materilizer as well

3. `ChangelogSnapshotState` and its updates are kept in ChangelogKeyedStatBackend. Materialization works as a stateless Materialization Manager providing function utilities.
    - Pros:
	  1. All states are wrapped in ChangelogKeyedStatBackend
	  2. Conceptually also works naturally
	  
    - Cons:
	 Circular constructor. `Materialization Manager` needs access to `ChangelogKeyedStatBackend` to update `ChangelogSnapshotState`
	 `ChangelogKeyedStatBackend` is created from StateBackend#createKeyedStateBackend. 
	  
	  To avoid circular construction, `Materialization Manager` has to be exposed at the time creating ChangelogKeyedStatBackend. 

@rkhachatryan what do you think Roman?

pnowojski commented:
@curcur what do you mean by `Materialization`? I don't see any interface/class named like that? Do you mean just the materialization process/trigger?

rkhachatryan commented:
Thanks a lot for the write-up Yuan!

To clarify my concerns regarding option 2:
States (e.g. `ChangelogValueState`) [connect](https://github.com/apache/flink/blob/7901f26d3470389c7fcfc720e2bab32cc226f1e6/flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java#L62) `StateChangelogWriter` with the underlying backend: each state change must be reflected atomically in both.
When creating a snapshot, mirroring this atomicity is crucial for correctness.
That's why I think it's better to keep these two parts as close to each other as possible (and states are part of the `ChangelogKeyedStateBackend`, while materializer is more independent).

In option 3, circular dependency is redundant if `triggerMaterialization` method is removed (that's what I did in the original commit: instead of triggering the materialization, test can [waits](https://github.com/rkhachatryan/flink/commit/61f37810ecccca53791547a052d4e895b8e4f513#diff-8654aaea028ca18fe809d05ce46a4a17cf3a4b18f72ec588f02611d47956eb19R5140) materiazation to happen). So it seems like this approach doesn't have major concerns.

curcur commented:
> @curcur what do you mean by `Materialization`? I don't see any interface/class named like that? Do you mean just the materialization process/trigger?

I've updated the comment with related links @pnowojski . Materialization means the general materialization process, including trigger (sync part), and uploading (async part).

curcur commented:
> Thanks a lot for the write-up Yuan!
> 
> To clarify my concerns regarding option 2:
> States (e.g. `ChangelogValueState`) [connect](https://github.com/apache/flink/blob/7901f26d3470389c7fcfc720e2bab32cc226f1e6/flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogValueState.java#L62) `StateChangelogWriter` with the underlying backend: each state change must be reflected atomically in both.
> When creating a snapshot, mirroring this atomicity is crucial for correctness.
> That's why I think it's better to keep these two parts as close to each other as possible (and states are part of the `ChangelogKeyedStateBackend`, while materializer is more independent).

Yes, this part I agree. The problem of put `ChangelogSnapshotState` in `PeriodicMaterializer` is that we have sync up in these two places to guarantee this atomicity


> In option 3, circular dependency is redundant if `triggerMaterialization` method is removed (that's what I did in the original commit: instead of triggering the materialization, test can [waits](https://github.com/rkhachatryan/flink/commit/61f37810ecccca53791547a052d4e895b8e4f513#diff-8654aaea028ca18fe809d05ce46a4a17cf3a4b18f72ec588f02611d47956eb19R5140) materiazation to happen). So it seems like this approach doesn't have major concerns.

This part I do not agree with. It is not about tests. 
As long as you have `ChangelogSnapshotState` and its update methods in `ChangelogKeyedStatebackend`, you will have the same problem. Unless you also put the periodic triggering logic into `ChangelogKeyedStatebackend`, I cannot see how you can avoid circular constructor. This way, you purely leave uploading logic in `PeriodicMaterializationManager`.

In which way, it is very similar to Option1.

rkhachatryan commented:
To replicate 15 Sep offline discussion:
> Unless you also put the periodic triggering logic into ChangelogKeyedStatebackend, I cannot see how you can avoid circular constructor. But if you do so, you purely leave uploading logic in PeriodicMaterializationManager.

In my [branch](https://github.com/rkhachatryan/flink/tree/f21357-separated) I've put the periodic triggering logic into MaterializationManager, and avoided circular constructor

curcur commented:
rebase on master

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/ChangelogOptions.java:
```diff
@@ -0,0 +1,88 @@
+import java.time.Duration;
+
+/** A collection of all configuration options that relate to changelog. */
+public class ChangelogOptions {
```
Could you please extract this change into a separate commit?

rkhachatryan replied:
WDYT about marking the class as `@PublicEvolving` and adding "state" to name, like `StateChangelogOptions`?

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/ChangelogOptions.java:
```diff
@@ -0,0 +1,88 @@
+public class ChangelogOptions {
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_CHANGELOG)
+    public static final ConfigOption<Duration> PERIODIC_MATERIALIZATION_INTERVAL =
+            ConfigOptions.key("state.backend.periodic-materialize.interval")
```
Add changelog prefix to key? (ditto other keys)
Maybe `state.backend.changelog.periodic-materialization.interval`?

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/ChangelogOptions.java:
```diff
@@ -0,0 +1,88 @@
+                    .withDescription(
+                            "Defines the interval in milliseconds to perform periodic materialization for state backend.");
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_CHANGELOG)
+    public static final ConfigOption<Boolean> PERIODIC_MATERIALIZATION_ENABLED =
+            ConfigOptions.key("state.backend.periodic-materialize.enabled")
```
Remove? (as discussed above)

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/ChangelogOptions.java:
```diff
@@ -0,0 +1,88 @@
+                            "Whether to perform periodic materialization for state backend.");
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_CHANGELOG)
+    public static final ConfigOption<Integer> MATERIALIZATION_MAX_ALLOWED_FAILURES =
+            ConfigOptions.key("state.backend.periodic-materialize.enabled")
```
Wrong key I guess?

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/AsyncExceptionHandler.java:
```diff
@@ -16,7 +16,7 @@
  */
 
-package org.apache.flink.streaming.runtime.tasks;
+package org.apache.flink.runtime.taskmanager;
```
Could you please extract this change (moving of `AsyncExceptionHandler`) into a separate commit?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -531,18 +539,76 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
+    public void updateChangelogSnapshotState(
+            SnapshotResult<KeyedStateHandle> materializedSnapshot, SequenceNumber upTo) {
+        changelogSnapshotState =
+                new ChangelogSnapshotState(
+                        getMaterializedResult(materializedSnapshot), new ArrayList<>(), upTo);
```
nit: `Collections.emptyList`?
nit: factory method with two args?

curcur replied:
> nit: factory method with two args?

what's that?

rkhachatryan replied:
I mean a method like this in `ChangelogSnapshotState` class:
```
        public static ChangelogSnapshotState materialized(
                SnapshotResult<KeyedStateHandle> snapshot, SequenceNumber upTo) {
            return new ChangelogSnapshotState(
                    getMaterializedResult(snapshot), Collections.emptyList(), upTo);
        }
```

curcur replied:
Hmm, I think that's too verbose to have two levels of duplicated code to express the same thing.

`updateChangelogSnapshotState` is quite light.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java:
```diff
@@ -625,4 +691,52 @@ public T get(long timeout, TimeUnit unit)
+                List<KeyedStateHandle> materializedSnapshot,
+                List<ChangelogStateHandle> restoredNonMaterialized,
+                SequenceNumber materializedTo) {
+            this.materializedSnapshot = checkNotNull(unmodifiableList((materializedSnapshot)));
+            this.restoredNonMaterialized = checkNotNull(unmodifiableList(restoredNonMaterialized));
```
Isn't null check already done in `unmodifiableList`?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,252 @@
+            long periodicMaterializeDelay,
+            int allowedNumberOfFailures,
+            ChangelogKeyedStateBackend<?> keyedStateBackend) {
+        this.mailboxExecutor = mailboxExecutor;
+        this.asyncOperationsThreadPool = asyncOperationsThreadPool;
```
Check for null (ditto other args)?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,252 @@
+                                "periodic-materialization-scheduler-" + subtaskName));
+    }
+
+    public void start() {
```
Can we prevent this method from being called twice?
Otherwise, we can get very strange behavior (in case of programmatic error and calling it twice).

curcur replied:
It is only called in `ChangelogStateBackend#restore`. Where do you find called in other places (it is called in tests for testing PeriodicMaterializationManager)

rkhachatryan replied:
It's not called twice currently - and I'm proposing to enforce this; so that it's not re-used for example.

curcur replied:
yes, that's a good idea!

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,252 @@
+            FileSystemSafetyNet.closeSafetyNetAndGuardedResourcesForThread();
+        }
+
+        // if succeed, update state and finish up
+        if (materializedSnapshot != null) {
```
The flow isn't very clear to follow here. How about extracting the code above into a function that creates this `materializedSnapshot`?

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,252 @@
+
+            final SnapshotResult<KeyedStateHandle> copyMaterializedSnapshot = materializedSnapshot;
+
+            mailboxExecutor.execute(
+                    () -> {
+                        keyedStateBackend.updateChangelogSnapshotState(
+                                copyMaterializedSnapshot, upTo);
+                        numberOfConsecutiveFailures.set(0);
```
Do we have update `numberOfConsecutiveFailures` in the task thread?

curcur replied:
Not have to itself

But clean->0 should happen after `updateChangelogSnapshotState `

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,252 @@
+        scheduleNextMaterialization();
+    }
+
+    private void handleExecutionException(
+            RunnableFuture<SnapshotResult<KeyedStateHandle>> materializedRunnableFuture) {
```
There is already error handling logic calling this method. How about renaming this method to `discardFailedUpload` or something like this ?

rkhachatryan reviewed (changes requested):
Thanks for updating the PR @curcur. The logic of production changes looks mostly good to me.
Could you please extract two refactoring commits (I've left comments about it)?

I didn't review the test as agreed offline.

curcur commented:
Thanks @rkhachatryan for review! yeah, I will split this to separate commits!

rkhachatryan commented on flink-core/src/main/java/org/apache/flink/configuration/ChangelogOptions.java:
```diff
@@ -0,0 +1,81 @@
+                                    + "periodic materialization for state backend.");
+
+    @Documentation.Section(Documentation.Sections.STATE_BACKEND_CHANGELOG)
+    public static final ConfigOption<Integer> MATERIALIZATION_MAX_FAILURES_ALLOWED =
+            ConfigOptions.key("state.backend.changelog.max.failures.allowed")
```
`state.backend.changelog.max-failures`
(I think dot is already used to separate sections here)

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/execution/Environment.java:
```diff
@@ -258,4 +259,10 @@ default void setAsyncOperationsThreadPool(ExecutorService executorService) {}
     default ExecutorService getAsyncOperationsThreadPool() {
         throw new UnsupportedOperationException();
     }
+
+    default void setCheckpointStorageAccess(CheckpointStorageAccess checkpointStorageAccess) {}
+
+    default CheckpointStorageAccess getCheckpointStorageAccess() {
+        throw new UnsupportedOperationException();
+    }
```
I think this (and related) change belong to the wrong commit.

curcur replied:
> Thanks updating the PR @curcur.
> 
> I've left some comments, PTAL.
> 
> Besides that, could you please name the extracted commits according to [conventions](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#commit-naming-conventions) (add `[hotfix]` and component label, fix case, etc.).

which of the commits do you think not following the convention?

If you mean the [hotfix] one, It is to fix a file naming issue for log4j2 file. I am not sure whether it is belonging to a "component"

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/test/resources/log4j2-test.properties:
```diff
@@ -18,11 +18,11 @@
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = INFO
```
The common approach is to disable logging by default and change level on demand. Can't we use it here?

ditto: pattern

curcur replied:
BTW, what's the problem for `layout.pattern`? Do you mean we should remove it?

rkhachatryan replied:
> I will change the logging level back.

Thanks!
> BTW, what's the problem for layout.pattern? Do you mean we should remove it?

Actually, the change is correct, non-standard pattern was used before, sorry. Let's keep this change.

rkhachatryan commented on flink-state-backends/flink-statebackend-changelog/src/test/java/org/apache/flink/state/changelog/ChangelogStateBackendTestUtils.java:
```diff
@@ -76,4 +120,145 @@ public static TestTaskStateManager createTaskStateManager(File changelogStorageP
+
+            keyedBackend.setCurrentKey(2);
+            state.update(new StateBackendTestBase.TestPojo("u2", 2));
+
+            Thread.sleep(100);
```
I'm assuming this part isn't finished yet. 
Otherwise, this will make the test unstable.

rkhachatryan reviewed (changes requested):
Thanks updating the PR @curcur.

I've left some comments, PTAL.

Besides that, could you please name the extracted commits according to [conventions](https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#commit-naming-conventions) (add `[hotfix]` and component label, fix case, etc.).

curcur commented on pom.xml:
```diff
@@ -1518,7 +1518,7 @@ under the License.
 							unset: don't alter the configuration
 						-->
-						<checkpointing.changelog>random</checkpointing.changelog>
+						<checkpointing.changelog>on</checkpointing.changelog>
```
yes, it will be changed to random before merging. It is used purely for testing

rkhachatryan replied:
Could you then move it to a separate commit please?

curcur commented on flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java:
```diff
@@ -0,0 +1,269 @@
+                subtaskName,
+                periodicMaterializeDelay / 1000);
+
+        scheduleNextMaterialization();
```
If the retry limit is exceeded, the task will be failed and trigger failover

Do you mean it is possible that before failover is triggered, another materialization is triggered?

curcur replied:
What's the problem if `scheduleNextMaterialization` is called?

rkhachatryan replied:
> Do you mean it is possible that before failover is triggered, another materialization is triggered?
> And what's the problem if scheduleNextMaterialization is called?

Yes. One problem is that it likely will fail (because some of the executors is shut/ting down),  and potentially prevent failover. If not, it's just unnecessary upload.

Besides that, the code would be much cleaner IMO.

curcur commented:
@rkhachatryan , I've removed the change ``waiting for termination of periodicExecutor''

The periodicExecutor is the executor to trigger the scheduling, I am not sure awaits it helps to solve the problem of untracked files (untracked files are uploaded by a separate thread pool). And since most of the time of a task is used for waiting the `periodicMaterializeDelay`, It has a very high probability that each shut down has to wait till timeout.

It does not bring too many benefits but make the shutdown of a flink job much slower than before. 

So I removed the awaits.
